src/common/sourceMaps/codeSearchSourceMapRepository.ts

```diff
@@ -0,0 +1,88 @@
+/*---------------------------------------------------------
+ * Copyright (C) Microsoft Corporation. All rights reserved.
+ *--------------------------------------------------------*/
+
+import { ISourceMapMetadata } from './sourceMap';
+import { ISourceMapRepository, createMetadataForFile } from './sourceMapRepository';
+import { NodeSourceMapRepository } from './nodeSourceMapRepository';
+import { logger } from '../logging/logger';
+import { LogTag } from '../logging';
+import { forceForwardSlashes } from '../pathUtils';
+
+type FindTextFn = (typeof import('vscode'))['workspace']['findTextInFiles'];
+
+/**
+ * A source map repository that uses VS Code's proposed search API to
+ * look for candidate files.
+ */
+export class CodeSearchSourceMapRepository implements ISourceMapRepository {
+  constructor(private readonly findFn: FindTextFn) {}
+
+  public static createOrFallback() {
+    try {
+      const code: typeof import('vscode') = require('vscode');
+      if (code.workspace.findTextInFiles) {
+        return new CodeSearchSourceMapRepository(
+          code.workspace.findTextInFiles.bind(code.workspace),
+        );
+      }
+    } catch {
+      // ignored -- VS won't have vscode as a viable import, fall back to the memory/node.js version
+    }
+
+    return new NodeSourceMapRepository();
+  }
+
+  /**
+   * Returns the sourcemaps in the directory, given as an absolute path..
+   */
+  public async findDirectChildren(
+    _absolutePath: string,
+  ): Promise<{ [path: string]: Required<ISourceMapMetadata> }> {
+    throw new Error('not implemented');
+  }
+
+  /**
+   * Recursively finds all children of the given direcotry.
+   */
+  public async streamAllChildren<T>(
+    patterns: ReadonlyArray<string>,
+    onChild: (child: Required<ISourceMapMetadata>) => Promise<T>,
+  ): Promise<T[]> {
+    const todo: Promise<T | void>[] = [];
+
+    await this.findFn(
+      { pattern: 'sourceMappingURL', isCaseSensitive: true },
+      {
+        // todo(connor4312): is this correct way to join globs for search?
+        include: forceForwardSlashes(patterns.filter(p => !p.startsWith('!')).join(', ')),
+        exclude:
+          patterns
+            .filter(p => p.startsWith('!'))
+            .map(p => forceForwardSlashes(p.slice(1)))
+            .join(','),
+        useIgnoreFiles: false,
```

Comment:
Personally I have vscode set up to ignore my build directory, since it just clutters things up and I usually never need to look in there, which would break prediction in this scenario. That is, the desire behind ignoring files in vscode may not be equivalent to the desire to ignore files for the purpose of breakpoint discovery. I think having the option to explicitly exclude folders via negated globs in the `outFiles` is probably fine...

Reply:
Yeah good point.